server/utils/availabilityHelper.js

```diff
@@ -0,0 +1,230 @@
+        // console.log('day fully busy, go to next day');
+        currDay.add(1, 'day');
+      } else {
+        userStart = moment(start.format()).hour(userAvail.hours.start.hour).minute(userAvail.hours.start.minute);
+        userEnd = moment(start.format()).hour(userAvail.hours.end.hour).minute(userAvail.hours.end.minute);
+        if (userStart.isBefore(start)) userStart.add(1, 'day');
+        // console.log(start.format(), start.isSameOrBefore(end), start.format('hhmm'), end.format());
+        x = recursIsDayAvail(freebusy, b, start.format(), end.format(), reqMeet, userStart, userEnd, userTz);
+        if (x[0]) avail.push(x[2]);
+        b = x[1];
+        currDay.add(1, 'day');
+      }
+    }
+  }
+
+  let clientAvail = avail.map((time) => moment.tz(time, clientTz).date());
+
+  clientAvail = clientAvail.filter((el, i) => clientAvail.indexOf(el) === i);
+  //remove duplicate dates if any
+  return clientAvail;
+}
+
+function addToSlots(currStart, availEnd, reqMeet, slotTime, slots) {
+  while (currStart.isBefore(availEnd)) {
+    if (Math.abs(currStart.diff(availEnd, 'minutes')) >= reqMeet) {
+      slots.push(currStart.format());
+      currStart.add(slotTime, 'm');
+    } else {
+      break;
+    }
+  }
+
+  return slots;
+}
+
+function availSlots(date, freebusy, userHours, userTz, clientTz, reqMeet) {
+  const slots = [];
+  const slotTime = reqMeet > 15 ? 30 : 15;
+  let b = 0;
+
+  //client date -> usertz -> set to userhours times
+  let busyStart, busyEnd, busy;
+  let curr = moment.tz(date, clientTz).tz(userTz);
+  const end = moment(curr.format()).add(1, 'day');
+  const userStart = moment(curr.format()).hour(userHours.start.hour).minute(userHours.start.minute);
+  const userEnd = moment(curr.format()).hour(userHours.end.hour).minute(userHours.end.minute);
+
+  const endFirst = userStart.isBefore(curr);
+  if (endFirst) userStart.add(1, 'day');
+
+  while (curr.isBefore(end)) {
+    if (Math.abs(curr.diff(end, 'minutes')) < reqMeet) {
+      break;
+    }
+    busy = freebusy[b];
+    if (busy) {
+      busyStart = moment(busy.start);
+      busyEnd = moment(busy.end);
+    }
+
+    if (busyEnd.isSameOrBefore(curr)) {
+      curr = busyEnd;
+      b++;
+    }
+    if (endFirst && curr.isBefore(userEnd)) {
+      if (busy && busyStart.isBefore(userEnd)) {
+        addToSlots(curr, busyStart, reqMeet, slotTime, slots);
+        curr = busyEnd;
+        b++;
+      } else {
+        addToSlots(curr, userEnd, reqMeet, slotTime, slots);
+        curr = userStart;
+      }
+    } else if (endFirst && curr.isSameOrAfter(userStart)) {
+      if (busy && busyStart.isBefore(end)) {
+        addToSlots(curr, busyStart, reqMeet, slotTime, slots);
+        curr = busyEnd;
+        b++;
+      } else {
+        addToSlots(curr, end, reqMeet, slotTime, slots);
+        curr = end;
+      }
+    } else if (!endFirst && curr.isBetween(userStart, userEnd, null, '[)')) {
```

Comment:
let's add some comments for each of these if statements explaining what it's comparing - especially for this comparison for `[}` :)